### DISCUSSION TITLE
Enable two authentication backends

### DISCUSSION THREAD

**Discussion by bert2002** [fastapi-users/fastapi-users#989] [General]
Hi,
before the latest updates I could enable two authentication backends (cookie and jwt), but I dont see how I can do it for the new method. At the moment I have:
```
SECRET = os.environ.get('SECRET')
auth_backends = []

# JWT AUTH
bearer_transport = BearerTransport(tokenUrl="auth/jwt/login")

def get_jwt_strategy() -> JWTStrategy:
  return JWTStrategy(secret=SECRET, lifetime_seconds=86400)

auth_backend = AuthenticationBackend(
  name="auth_jtw",
  transport=bearer_transport,
  get_strategy=get_jwt_strategy,
)

# COOKIE AUTH
cookie_transport = CookieTransport(cookie_max_age=86400)

auth_backend = AuthenticationBackend(
  name="auth",
  transport=cookie_transport,
  get_strategy=get_jwt_strategy,
)
```
Even when I call `auth/jwt/login` it will respond with the cookie and not with the jwt token. When I disable the transport and backend for cookie auth, then it works. 

Is there a way to enable both again?

Thanks,
bert

**Comment by frankie567** [comment]
Hi @bert2002!

You're almost there actually 🙂 You have two authentication backends, but you should name the variables differently so you can identify them correctly in the code:

```py
# JWT AUTH
bearer_transport = BearerTransport(tokenUrl="auth/jwt/login")

def get_jwt_strategy() -> JWTStrategy:
  return JWTStrategy(secret=SECRET, lifetime_seconds=86400)

jwt_auth_backend = AuthenticationBackend(
  name="jwt",   # I changed the name
  transport=bearer_transport,
  get_strategy=get_jwt_strategy,
)

# COOKIE AUTH
cookie_transport = CookieTransport(cookie_max_age=86400)

cookie_auth_backend = AuthenticationBackend(
  name="cookie",  # I changed the name
  transport=cookie_transport,
  get_strategy=get_jwt_strategy,
)
```

Then, you need to add them both to the `FastAPIUsers` initializer. This will make sure the `current_user` dependency will try to check for a token in both backends:
 
```py
fastapi_users = FastAPIUsers[User, uuid.UUID](
    get_user_manager,
    [jwt_auth_backend, cookie_auth_backend],
)
```

Finally, we need a way for a user to authenticate either with a JWT Bearer or a JWT Cookie. For this, you need to create two distinct auth routers, each one with a different backend:

```py
app.include_router(
    fastapi_users.get_auth_router(jwt_auth_backend),
    prefix="/auth/jwt",
    tags=["auth"],
)

app.include_router(
    fastapi_users.get_auth_router(cookie_auth_backend),
    prefix="/auth/cookie",
    tags=["auth"],
)
```

And that's it  🚀

**Reply by bert2002** [reply]
That indeed works very well. Thanks a lot, I missed that I can declare different variables. Might be worth adding it to the documentation.
